File: java/src/main/java/com/sketches_oxide/membership/BlockedBloomFilter.java
package com.sketches_oxide.membership;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * BlockedBloomFilter - Cache-Efficient Probabilistic Membership Testing.
 *
 * A cache-optimized variant of the classic Bloom filter that organizes bits into
 * cache-line-sized blocks. All hash functions for an item map to the same block,
 * resulting in much better cache locality and significantly faster performance
 * on modern CPUs.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Divides the bit array into blocks (typically 64 bytes / 512 bits)</li>
 *   <li>Each item hashes to exactly one block</li>
 *   <li>All k hash functions set/check bits within that single block</li>
 *   <li>Minimizes cache misses: 1 cache line access vs k random accesses</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>High-throughput duplicate detection</li>
 *   <li>Database query optimization (avoiding disk seeks)</li>
 *   <li>Network packet filtering</li>
 *   <li>Any application where Bloom filter speed is critical</li>
 * </ul>
 *
 * <p><strong>Performance vs Standard Bloom:</strong>
 * <ul>
 *   <li>Speed: 2-4x faster on typical workloads</li>
 *   <li>Space: Slightly larger (~10-20%) for same FPR</li>
 *   <li>FPR: Same theoretical guarantees</li>
 *   <li>Cache efficiency: O(1) cache misses vs O(k)</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (BlockedBloomFilter bf = new BlockedBloomFilter(1_000_000, 0.01)) {
 *     // Bulk insert phase
 *     for (String url : urls) {
 *         bf.insert(url.getBytes());
 *     }
 *
 *     // Fast lookup phase
 *     if (bf.contains("http://example.com".getBytes())) {
 *         System.out.println("URL might be in the set");
 *     }
 * }
 * </pre>
 *
 * @see BloomFilter
 * @see <a href="https://algo.inria.fr/flajolet/Publications/PuFl05.pdf">Cache-Conscious Bloom Filters</a>
 */
public final class BlockedBloomFilter extends NativeSketch implements MergeableSketch<BlockedBloomFilter> {

    private final long n;
    private final double fpr;

    /**
     * Create a new BlockedBloomFilter.
     *
     * @param n expected number of items to insert
     * @param fpr desired false positive rate (0 &lt; fpr &lt; 1)
     *            Common values: 0.01 (1%), 0.001 (0.1%), 0.0001 (0.01%)
     * @throws IllegalArgumentException if n &lt;= 0 or fpr not in (0, 1)
     * @throws RuntimeException if native memory allocation fails
     */
    public BlockedBloomFilter(long n, double fpr) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive, got: " + n);
        }
        if (fpr <= 0 || fpr >= 1) {
            throw new IllegalArgumentException("fpr must be in range (0, 1), got: " + fpr);
        }

        this.n = n;
        this.fpr = fpr;
        this.nativePtr = SketchOxideNative.blockedbloomfilter_new(n, fpr);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate BlockedBloomFilter");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private BlockedBloomFilter(long nativePtr, long n, double fpr) {
        this.nativePtr = nativePtr;
        this.n = n;
        this.fpr = fpr;
    }

    /**
     * Insert an item into the filter.
     *
     * The item's hash is computed and mapped to a single cache block,
     * where all k bits are set. This operation is typically 2-4x faster
     * than standard Bloom filter insertion.
     *
     * @param item the item to insert (as raw bytes)
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.blockedbloomfilter_insert(nativePtr, item);
    }

    /**
     * Insert a string item into the filter.
     *
     * Convenience method that converts the string to UTF-8 bytes.
     *
     * @param item the string to insert
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        insert(item.getBytes());
    }

    /**
     * Check if an item might be in the filter.
     *
     * Returns true if the item might be in the set (with false positive rate = fpr).
     * Returns false if the item is definitely not in the set (no false negatives).
     *
     * All k bits are checked within a single cache block, making this operation
     * significantly faster than standard Bloom filter queries.
     *
     * @param item the item to check (as raw bytes)
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.blockedbloomfilter_contains(nativePtr, item);
    }

    /**
     * Check if a string item might be in the filter.
     *
     * Convenience method that converts the string to UTF-8 bytes.
     *
     * @param item the string to check
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return contains(item.getBytes());
    }

    /**
     * Merge another BlockedBloomFilter into this one.
     *
     * After merging, this filter contains all items from both filters.
     * Both filters must have the same n and fpr parameters.
     *
     * @param other the filter to merge into this one
     * @throws IllegalStateException if either filter is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(BlockedBloomFilter other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.n != other.n || Math.abs(this.fpr - other.fpr) > 1e-9) {
            throw new IllegalArgumentException(
                    "Cannot merge BlockedBloomFilters with different parameters: " +
                            String.format("(n=%d, fpr=%.6f) vs (n=%d, fpr=%.6f)",
                                    this.n, this.fpr, other.n, other.fpr));
        }

        SketchOxideNative.blockedbloomfilter_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Serialize the filter to binary format.
     *
     * The serialized format includes all configuration and bit array data,
     * allowing full reconstruction of the filter state.
     *
     * @return binary representation of the filter
     * @throws IllegalStateException if the filter has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.blockedbloomfilter_serialize(nativePtr);
    }

    /**
     * Deserialize a BlockedBloomFilter from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new BlockedBloomFilter instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static BlockedBloomFilter deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.blockedbloomfilter_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized BlockedBloomFilter data");
        }

        // Create with dummy values - actual values are in native struct
        return new BlockedBloomFilter(ptr, 100, 0.01);
    }

    /**
     * Get the expected number of items.
     *
     * @return the n parameter used at creation
     */
    public long getN() {
        return n;
    }

    /**
     * Get the target false positive rate.
     *
     * @return the fpr parameter used at creation
     */
    public double getFpr() {
        return fpr;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.blockedbloomfilter_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("BlockedBloomFilter(n=%d, fpr=%.6f)", n, fpr);
        }
        return "BlockedBloomFilter(closed)";
    }
}

File: dotnet/SketchOxide/src/Membership/BlockedBloomFilter.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Membership;

/// <summary>
/// Cache-optimized Blocked Bloom Filter for probabilistic set membership testing.
///
/// A variant of Bloom filter that groups bits into cache-line-sized blocks,
/// improving cache locality and memory access patterns. This results in
/// significantly faster lookups, especially for large filters.
/// </summary>
/// <remarks>
/// <para>
/// The Blocked Bloom filter provides:
/// - Better cache performance than standard Bloom filter
/// - Single cache-line access per lookup
/// - Slightly higher false positive rate than standard Bloom filter
/// - Same memory usage as standard Bloom filter
/// </para>
/// <para>
/// Ideal for applications where:
/// - Lookup performance is critical
/// - Filter size exceeds CPU cache
/// - High throughput is required
/// </para>
/// </remarks>
public sealed class BlockedBloomFilter : NativeSketch
{
    private readonly ulong _size;
    private readonly double _fpr;

    /// <summary>
    /// Creates a new Blocked Bloom filter with the specified expected number of elements and false positive rate.
    /// </summary>
    /// <param name="expectedElements">Expected number of elements to be inserted. Must be greater than 0.</param>
    /// <param name="falsePositiveRate">Desired false positive rate (e.g., 0.01 for 1%). Must be in range (0, 1).</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if expectedElements is 0 or falsePositiveRate is outside (0, 1).</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public BlockedBloomFilter(ulong expectedElements, double falsePositiveRate)
    {
        if (expectedElements == 0)
            throw new ArgumentOutOfRangeException(nameof(expectedElements), expectedElements, "Expected elements must be greater than 0");
        if (falsePositiveRate <= 0 || falsePositiveRate >= 1)
            throw new ArgumentOutOfRangeException(nameof(falsePositiveRate), falsePositiveRate, "False positive rate must be in range (0, 1)");

        _size = expectedElements;
        _fpr = falsePositiveRate;
        NativePtr = SketchOxideNative.blockedbloomfilter_new(expectedElements, falsePositiveRate);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native BlockedBloomFilter");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private BlockedBloomFilter(ulong size, double fpr, nuint ptr)
    {
        _size = size;
        _fpr = fpr;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the expected number of elements this filter was configured for.
    /// </summary>
    public ulong ExpectedElements
    {
        get
        {
            CheckAlive();
            return _size;
        }
    }

    /// <summary>
    /// Gets the target false positive rate.
    /// </summary>
    public double FalsePositiveRate
    {
        get
        {
            CheckAlive();
            return _fpr;
        }
    }

    /// <summary>
    /// Inserts an element into the filter.
    /// </summary>
    /// <param name="data">The bytes representing the element to insert.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Insert(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.blockedbloomfilter_insert(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Inserts a string element into the filter.
    /// </summary>
    /// <param name="value">The string to insert.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Insert(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Insert(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Tests whether an element might be in the set.
    /// </summary>
    /// <param name="data">The bytes representing the element to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.blockedbloomfilter_contains(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Tests whether a string element might be in the set.
    /// </summary>
    /// <param name="value">The string to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Contains(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the filter to a byte array.
    /// </summary>
    /// <returns>Serialized filter bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.blockedbloomfilter_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a Blocked Bloom filter from a byte array.
    /// </summary>
    /// <param name="data">Serialized filter bytes.</param>
    /// <param name="expectedElements">The expected elements parameter used when creating the original filter.</param>
    /// <param name="falsePositiveRate">The false positive rate parameter used when creating the original filter.</param>
    /// <returns>A new BlockedBloomFilter instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static BlockedBloomFilter Deserialize(byte[] data, ulong expectedElements, double falsePositiveRate)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.blockedbloomfilter_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize BlockedBloomFilter: invalid data");

        return new BlockedBloomFilter(expectedElements, falsePositiveRate, ptr);
    }

    /// <summary>
    /// Returns a string representation of the filter.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "BlockedBloomFilter(disposed)";
        return $"BlockedBloomFilter(expectedElements={_size}, fpr={_fpr:P2})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.blockedbloomfilter_free(NativePtr);
            NativePtr = 0;
        }
    }
}
